### ISSUE TITLE
 Word2Vec.load_word2vec_format should support python file like objects

### ISSUE THREAD

**Issue by pgroth** [piskvorky/gensim#372]
Currently, the loading code requires a filename. It would be nice to allow a file object to be given instead. This would make it easier to load models in environments (e.g. Spark) that don't necessarily have access to a normal file system.

**Comment by piskvorky** [comment]
Just use pickle manually then.

The point of `save` / `load` is they do some extra work for efficiency (storing large arrays in separate files, memory mapping, transparent compression).

None of that is possible with just a file descriptor, so you might as well pickle the object manually.

CC @gojomo @cscorley thoughts?

**Comment by pgroth** [comment]
Ok. Maybe I'll look at what happens when I call open directly on an hdfs
path.



Thanks!
Paul

**Comment by gojomo** [comment]
There was a comment somewhere that pickle might encounter problems on very-large numpy arrays. Is that true? If so, it'd be good to ensure there's a workaround for that other than the multi-path filesystem-persistence, since it's always nice to have the option to load/save from arbitrary streams/pipe/file-likes.

**Comment by piskvorky** [comment]
Yes, that's one reason why gensim has a dedicated `save`. Memory mapping is the other.

Some HDF5-like options could be interesting too. Let me know if you write something in that direction, we could include that in gensim too (if it solves a common enough use-case with a sane enough API).

**Comment by piskvorky** [comment]
Oops, I only just noticed this ticket is about `load_word2vec_format`, not normal `save`/`load`.

@pgroth that's a much easier proposition -- can you submit a PR that will allow file-like param in `load_word2vec_format`? Maybe we can just overload the same param, and check whether it's to be interpreted as filename or not (=string or not).

By the way, allowing file handles in `save`/`load` was added in #292 (gensim 0.11.0), though I haven't used it much myself.

**Comment by pgroth** [comment]
Sure. I'll probably do that next week or so.

**Comment by wikier** [comment]
What's the status of this?

**Comment by petabyte** [comment]
was this resolved? what version?
This is is what I get when I pass a file object to load
  File "Lib\site-packages\gensim\models\word2vec.py", line 1470, in load
  File "Lib\site-packages\gensim\utils.py", line 264, in load
  File "Lib\site-packages\gensim\utils.py", line 329, in _adapt_by_suffix
AttributeError: 'file' object has no attribute 'endswith'

**Comment by pgroth** [comment]
No I don't think so. I never submitted a pull request

**Comment by yupbank** [comment]
hey, i also run into the same problem..  I'm loading a very big model (~1.6 Gb) from hdfs in gz format
and `smart_open` doesn't provide `readline` for hdfs files apparently.. what is the recommended way to deal with this?

**Comment by yupbank** [comment]
so i figured, the problem is that
[ ]: `smart_open('hdfs://xx.gz')` is not supported 
[ ]:  `smart_open('hdfs://xx').readline()` is not supported or don't need to, but the `load_word2vec_format` need to support file obj with `read` only.

**Comment by piskvorky** [comment]
@yupbank adding support for these two features (`readline()` and `gz` for HDFS) would be much appreciated!

**Comment by bhanu546** [comment]
Is anyone working on this as i can pick this up. The same issue persists with `Word2Vec.load` for loading model too as both uses the same method to load files.

**Comment by bhanu546** [comment]
Okay,i'd like to pick it up then, supporting file objects as an input along with filenames.

**Comment by menshikh-iv** [comment]
@bhanu546 feel free to pick

**Comment by halflings** [comment]
With 3.6 introducing the file-based training, this change would really be welcome.
Additional anecdote on my side: I can't use the file-based training because I can't make a full local copy of some data located in a data center; can only iterate over it (but this means I still need to use the queue-based system).
